airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/StateMessageHelper.java

```diff
@@ -0,0 +1,30 @@
+package io.airbyte.protocol.models;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.json.Jsons;
+import io.vavr.control.Either;
+import java.util.List;
+
+public class StateMessageHelper {
+
+  public static class AirbyteStateMessageListTypeReference extends TypeReference<List<AirbyteStateMessage>> {
+  }
+
+  /**
+   * This a takes a json blob state and tries return either a legacy state in the format of a json object or a state message with the new format
+   * which is a list of airbyte state message.
+   * @param state
+   * @return Either a json blob (on the left) or a structure state message.
+   */
+  public static Either<JsonNode, List<AirbyteStateMessage>> getTypeState(JsonNode state) {
```

Comment:
I return an Either instead of a list of state message. It could be a list with a single element but I wanted to potentially minimize the changes in a connector in order to clearly split the legacy from the non-legacy.

Reply:
in my implementation it allows me to do
```
Either<JsonNode, List<AirbyteStateMessage>> maybeState = StateMessageHelper.getTypeState(sourceConfig.getState().getState());
          if (maybeState.isLeft()) {
            log.error("The state is not compatible with a partial reset that  have been requested");
            throw new IllegalStateException("Legacy state for a partial reset");
          }
```

Reply:
@benmoriceau I think that it will be fine.  I'll have to adjust some code on the connector side, but should be easy enough.